client.py

```
@@ -0,0 +1,561 @@
+        try:
+            self.client.connect()
+            self.base_namespace_idx = self.client.get_namespace_index(self.base_namespace)
+            logger.info(f"Connected to OPC UA server at {self.endpoint}")
+            logger.debug(f"Base namespace '{self.base_namespace}' has index: {self.base_namespace_idx}")
+
+            namespaces = self.client.get_namespace_array()
+            for idx, ns in enumerate(namespaces):
+                if ns.startswith(f"{self.base_namespace}/"):
+                    device_name = ns.split("/")[-1]
+                    self.device_namespaces[device_name] = idx
+
+            return True
+        except Exception as e:
+            logger.error(f"Failed to connect to server: {e}")
+            return False
+
+    def disconnect(self):
+        """Disconnect from the OPC UA server"""
+        if self.client:
+            self.client.disconnect()
+            logger.info("Disconnected from OPC UA server")
+
+    def get_device_namespace_idx(self, device_name):
+        """Get the namespace index for a specific device"""
+        if device_name in self.device_namespaces:
+            return self.device_namespaces[device_name]
+
+        # If not cached, try to find it
+        try:
+            device_namespace = f"{self.base_namespace}/{device_name}"
+            idx = self.client.get_namespace_index(device_namespace)
+            self.device_namespaces[device_name] = idx
+            return idx
+        except Exception as e:
+            logger.error(f"Error getting namespace index for device {device_name}: {e}")
+            return None
+
+    def press_virtual_button(self, device_name):
+        """Press a virtual button"""
+        try:
+            button_node = self._get_node(device_name, "button", STATE_REGISTER)
+            if button_node:
+                button_node.set_value(True)
+                logger.info(f"Pressed virtual button: {device_name}")
+                return True
```

Again, i'll rather raise an exception than returning bools around. The exceptions can be typed and you could have control over different errors you could have